File: src/data/courses/dsa/binary-search-trees/views-vertical-problems/top-view/code/python/top_view.py
from collections import deque
class Solution:
    def topView(self, root):
        if not root: return []
        hd_map = {}
        queue = deque([(root, 0)])
        while queue:
            node, hd = queue.popleft()
            if hd not in hd_map: hd_map[hd] = node.val
            if node.left: queue.append((node.left, hd-1))
            if node.right: queue.append((node.right, hd+1))
        return [hd_map[k] for k in sorted(hd_map)]

File: src/data/courses/dsa/binary-search-trees/views-vertical-problems/top-view/code/java/TopView.java
import java.util.*;
class Solution {
    public List<Integer> topView(TreeNode root) {
        if (root==null) return new ArrayList<>();
        TreeMap<Integer,Integer> map = new TreeMap<>();
        Queue<Object[]> q = new LinkedList<>(); q.add(new Object[]{root,0});
        while (!q.isEmpty()) { Object[] c = q.poll(); TreeNode n=(TreeNode)c[0]; int hd=(int)c[1];
            map.putIfAbsent(hd, n.val);
            if(n.left!=null) q.add(new Object[]{n.left,hd-1});
            if(n.right!=null) q.add(new Object[]{n.right,hd+1}); }
        return new ArrayList<>(map.values());
    }
}